src/ComponentProvider.ts

```diff
@@ -1,94 +1,60 @@
+    '$.channels.*.*.message.traits[*]..[?(@.type)]',
+    '$.channels.*.*.message.headers',
+    '$.channels.*.*.message.headers..[?(@.type)]',
+    '$.channels.*.*.message.payload',
+    '$.channels.*.*.message.payload..[?(@.type)]',
+    '$.channels.*.parameters.*.schema[?(@.type)]',
+    '$.channels.*.parameters.*.schema..[?(@.type)]',
+    '$.components.schemas..[?(@.type)]',
+  ];
+
+  parameterPaths = ['$.channels.*.parameters.*', '$.components.parameters.*'];
   messages = new Map<string, Message>();
   schemas = new Map<string, Schema>();
   parameters = new Map<string, ChannelParameter>();
 
   constructor(private document: AsyncAPIDocument) {
-    this.scanChannels();
-    this.scanComponents();
-  }
-
-  private scanChannels(): void {
-    const channels = this.document.channels();
-    for (const channelName in channels) {
-      const path = `channels.${channelName}`;
-      const channel = this.document.channel(channelName);
-
-      if (channel.hasPublish()) {
-        const message = channel.publish().message();
-        const currentPath = `${path}.publish.message`;
-        this.scanMessage(currentPath, message);
-        this.messages.set(currentPath, message);
-      }
-
-      if (channel.hasSubscribe()) {
-        const message = channel.subscribe().message();
-        const currentPath = `${path}.subscribe.message`;
-        this.scanMessage(currentPath, message);
-        this.messages.set(currentPath, message);
-      }
-
-      if (channel.hasParameters()) {
-        const currentPath = `${path}.parameters`;
-        for (const [name, channelParameter] of Object.entries(channel.parameters())) {
-          this.parameters.set(`${currentPath}.${name}`, channelParameter);
-        }
-      }
-    }
-  }
-
-  private scanSchema(path: string, schema: Schema): void {
-    if (!schema) {return;}
-    this.schemas.set(path, schema);
-    const schemaProperties = schema.properties();
-    for (const [propertyName, propertySchema] of Object.entries(schemaProperties)) {
-      this.schemas.set(`${path}.properties.${propertyName}`, propertySchema);
-    }
-  }
-
-  private scanMessageTraits(path: string, traits: MessageTrait[]): void {
-    for (const [index, trait] of Object.entries(traits)) {
-      if (trait.headers()) {
-        this.scanSchema(`${path}[${index}].headers`, trait.headers());
-      }
-    }
+    this.messages = this.parseComponents(this.messagePaths);
+    this.schemas = this.parseComponents(this.schemaPaths);
+    this.parameters = this.parseComponents(this.parameterPaths);
   }
 
-  private scanMessage(path: string, message: Message): void {
-    this.scanSchema(`${path}.payload`, message.payload());
-    this.scanSchema(`${path}.headers`, message.headers());
-    this.scanMessageTraits(`${path}.traits`, message.traits());
+  private toLodashPath(JSONPath: string) {
+    return JSONPath.replace(/'\]\['/g, '.')
+      .slice(3, -2)
+      .replace(/'\]/g, '')
+      .replace(/\['/g, '.');
   }
```

Comment:
You should not name the argument as imported value. Shadowing of variables isn't good:

```suggestion
  private toLodashPath(path: string) {
    return path.replace(/'\]\['/g, '.')
      .slice(3, -2)
      .replace(/'\]/g, '')
      .replace(/\['/g, '.');
  }
```